amaranth_soc/csr/reg.py

```diff
@@ -590,319 +591,250 @@ def elaborate(self, platform):
-            try:
-                return cluster.get_path(register, _path=(*_path, name))
-            except KeyError:
-                pass
-
-        raise KeyError(register)
+        if not (isinstance(name, str) and name):
+            raise TypeError(f"Cluster name must be a non-empty string, not {name!r}")
+        self._scope_stack.append(name)
+        try:
+            yield
+        finally:
+            assert self._scope_stack.pop() == name
 
-    def get_register(self, path):
-        """Get a register.
+    @contextmanager
+    def Index(self, index):
+        """Define an array index.
 
         Arguments
         ---------
-        path : iter(:class:`str`)
-            Path of the register. It contains its name, prefixed by the name of parent clusters up
-            to this register map.
-
-        Returns
-        -------
-        :class:`Register`
-            The register assigned to ``path``.
+        index : :class:`int`
+            Array index.
 
         Raises
         ------
-        :exc:`ValueError`
-            If ``path`` is empty.
         :exc:`TypeError`
-            If ``path`` is not composed of non-empty strings.
-        :exc:`KeyError`
-            If ``path`` is not assigned to a register.
+            If ``index`` is not an integer, or is negative.
         """
-        path = tuple(path)
-        if not path:
-            raise ValueError("Path must be a non-empty iterable")
-        for name in path:
-            if not isinstance(name, str) or not name:
-                raise TypeError(f"Path must contain non-empty strings, not {name!r}")
-
-        name, *rest = path
-
-        if name in self._namespace:
-            assignment = self._namespace[name]
-            if not rest:
-                assert id(assignment) in self._registers
-                return assignment
+        if not (isinstance(index, int) and index >= 0):
+            raise TypeError(f"Array index must be a non-negative integer, not {index!r}")
+        self._scope_stack.append(index)
+        try:
+            yield
+        finally:
+            assert self._scope_stack.pop() == index
+
+    def as_memory_map(self):
+        self.freeze()
+        memory_map = MemoryMap(addr_width=self.addr_width, data_width=self.data_width,
+                               name=self.name)
+        for reg, reg_name, reg_offset in self._registers.values():
+            if reg_offset is not None:
+                reg_addr = (reg_offset * self.granularity) // self.data_width
             else:
-                assert id(assignment) in self._clusters
-                try:
-                    return assignment.get_register(rest)
-                except KeyError:
-                    pass
-
-        raise KeyError(path)
+                reg_addr = None
+            reg_size = (reg.element.width + self.data_width - 1) // self.data_width
+            # TBD: should integers be allowed inside resource names?
```

Comment:
Definitely. `lib.wiring` uses integers in paths, so SoC should be doing the same.

Reply:
https://github.com/amaranth-lang/amaranth-soc/issues/69